teuthology/openstack/__init__.py

```diff
@@ -0,0 +1,487 @@
+
+    def image_create(self, name):
+        """
+        Upload an image into OpenStack with glance. The image has to be qcow2.
+        """
+        misc.sh("wget -c -O " + name + ".qcow2 " + self.image2url[name])
+        misc.sh("glance image-create --property owner=teuthology " +
+                " --disk-format=qcow2 --container-format=bare " +
+                " --file " + name + ".qcow2 --name " + self.image_name(name))
+
+    def image(self, os_type, os_version):
+        """
+        Return the image name for the given os_type and os_version. If the image
+        does not exist it will be created.
+        """
+        name = self.type_version(os_type, os_version)
+        if not self.image_exists(name):
+            self.image_create(name)
+        return self.image_name(name)
+
+    def flavor(self, hint, select):
+        """
+        Return the smallest flavor that satisfies the desired size.
+        """
+        flavors_string = misc.sh("openstack flavor list -f json")
+        flavors = json.loads(flavors_string)
+        found = []
+        for flavor in flavors:
+            if select and not re.match(select, flavor['Name']):
+                continue
+            if (flavor['RAM'] >= hint['ram'] and
+                    flavor['VCPUs'] >= hint['cpus'] and
+                    flavor['Disk'] >= hint['disk']):
+                found.append(flavor)
+        if not found:
+            raise Exception("openstack flavor list: " + flavors_string +
+                            " does not contain a flavor in which" +
+                            " the desired " + str(hint) + " can fit")
+
+        def sort_flavor(a, b):
+            return (a['VCPUs'] - b['VCPUs'] or
+                    a['RAM'] - b['RAM'] or
+                    a['Disk'] - b['Disk'])
+        sorted_flavor = sorted(found, cmp=sort_flavor)
+        log.info("sorted flavor = " + str(sorted_flavor))
+        return sorted_flavor[0]['Name']
+
+    def cloud_init_wait(self, name_or_ip):
+        """
+        Wait for cloud-init to complete on the name_or_ip OpenStack instance.
+        """
+        log.debug('cloud_init_wait')
+        client_args = {
+            'timeout': 10,
+            'username': self.username,
+        }
+        if self.key_filename:
+            log.debug("using key " + self.key_filename)
+            client_args['key_filename'] = self.key_filename
+        with safe_while(sleep=2, tries=200,
+                        action="cloud_init_wait " + name_or_ip) as proceed:
+            success = False
+            # CentOS 6.6 logs in /var/log/clout-init-output.log
+            # CentOS 7.0 logs in /var/log/clout-init.log
+            all_done = ("tail /var/log/cloud-init*.log ; " +
+                        " test -f /tmp/init.out && tail /tmp/init.out ; " +
+                        " grep '" + self.up_string + "' " +
+                        "/var/log/cloud-init*.log")
+            while proceed():
+                client = paramiko.SSHClient()
```

Comment:
Curious why you're not using ``Remote`` here - is it just inappropriate?

It does seem like this ``while`` block might want to be split out as a psuedo-private method.